crates/sui-core/src/execution_cache/cache_types.rs

```diff
@@ -159,36 +183,121 @@ where
     pub fn new(cache_size: u64) -> Self {
         Self {
             cache: MokaCache::builder().max_capacity(cache_size).build(),
+            key_generation: (0..KEY_GENERATION_SIZE)
+                .map(|_| AtomicU64::new(0))
+                .collect(),
         }
     }
 
     pub fn get(&self, key: &K) -> Option<Arc<Mutex<V>>> {
         self.cache.get(key)
     }
 
+    fn generation(&self, key: &K) -> &AtomicU64 {
+        let mut state = DefaultHasher::new();
+        key.hash(&mut state);
+        let hash = state.finish();
+        &self.key_generation[(hash % KEY_GENERATION_SIZE as u64) as usize]
+    }
+
+    /// Get a ticket for caching the result of a read operation. The ticket will be
+    /// expired if a writer writes a new version of the value.
+    /// The caller must obtain the ticket BEFORE checking the dirty set and db. By
+    /// obeying this rule, the caller can be sure that if their ticket remains valid
+    /// at insert time, they either are inserting the most recent value, or a concurrent
+    /// writer will shortly overwrite their value.
+    pub fn get_ticket_for_read(&self, key: &K) -> Ticket {
+        let gen = self.generation(key);
+        Ticket::Read(gen.load(std::sync::atomic::Ordering::Acquire))
+    }
+
     // Update the cache with guaranteed monotonicity. That is, if there are N
     // calls to the this function from N threads, the write with the newest value will
     // win the race regardless of what ordering the writes occur in.
     //
     // Caller should log the insert with trace! and increment the appropriate metric.
-    pub fn insert(&self, key: &K, value: V) {
+    pub fn insert(&self, key: &K, value: V, ticket: Ticket) -> Result<(), ()> {
+        let gen = self.generation(key);
+
+        // invalidate other readers as early as possible. If a reader acquires a
+        // new ticket after this point, then it will read the new value from
+        // the dirty set (or db).
+        if matches!(ticket, Ticket::Write) {
+            gen.fetch_add(1, std::sync::atomic::Ordering::Release);
```

Comment:
Is this scenario possible?

Key 1 Gen 0 is not in cache

Writer calls insert Key 1 and fetch_add Gen 1
Reader obtains ticket Gen 1 for Key 1
Reader misses cache read and reads old value for Key 1 from DB
Writer writes new value for Key 1 to dirty cache
Cache value for Key 1 gets evicted
Reader caches old value in cache with valid Gen 1 ticket

If so, would swapping the generation fetch_add to after the cache update fix it? Or maybe cache eviction also updates generation?

Reply:
this is not possible because the writer holds the lock on the dirty cache while it calls fetch_add. reader waits for that lock before checking in the dirty cache. So reader will either complete its read before the fetch_add (in which case it will get an old ticket), or it will wait until after the fetch_add, in which case it is guaranteed to read the newest value from the dirty set (or possibly the db if the reader waits so long that the value has been flushed to the db).